tests/test-dladdr.c

```diff
@@ -0,0 +1,180 @@
+    int result;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return 2;
+    }
+
+    addr = dlsym( library, sym );
+    if ( !addr ) {
+        fprintf( stderr, "could not get address from library '%s' for symbol '%s'\n", libname, sym );
+        return 3;
+    }
+
+    result = check_dladdr( hint, addr, sym, should_match );
+    dlclose( library );
+
+    return result;
+}
+#endif
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#elif defined(_MSC_VER)
+/* disable warning C4273 inconsistent dll linkage */
+#pragma warning(push)
+#pragma warning(disable: 4273)
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+/* WINBASEAPI, which is normally used here, can be overriden by compiler or application, so we cannot use it */
+__declspec(dllimport) HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);
+__declspec(dllimport) int __cdecl atoi(const char *_Str);
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+#endif
+
+#ifdef _WIN32
+#define PassWithDifferentAddressOnWin PassWithDifferentAddress
+#define FailOnWin Fail
+#else
+#define PassWithDifferentAddressOnWin Pass
+#define FailOnWin Pass
+#endif
+
+#if defined(_WIN32) && !defined(DLFCN_WIN32_SHARED)
+#define PassOnSharedBuild Fail
+#else
+#define PassOnSharedBuild Pass
+#endif
+
+#define UNUSED(x) (void)(x)
+
+int main(int argc, char **argv)
+{
+    UNUSED(argv);
+    unsigned char buffer[6] = { 0xFF, 0x25, 0x00, 0x00, 0x00, 0x00 };
+    int  result = 0;
+
+    if (argc == 2)
+        verbose = 1;
+
+    result = check_dladdr( "null pointer", (void*)0, NULL , NoInfo);
+    result |= check_dladdr( "invalid pointer", (void*)0x125, NULL , NoInfo);
+    result |= check_dladdr( "function from dl library", ((void*)dladdr), "dladdr" , PassOnSharedBuild );
+    result |= check_dladdr( "function from dl library", (void*)dlopen, "dlopen", PassOnSharedBuild );
+    result |= check_dladdr( "function from glibc/msvcrt library", (void*)atoi, "atoi", Pass );
+    result |= check_dladdr( "function from executable", (void*)main, "main", FailOnWin );
+    result |= check_dladdr( "static function from executable", (void*)print_dl_info, "print_dl_info", Fail );
+    result |= check_dladdr( "address with negative offset", (char*)dladdr-6, "", Fail );
```

Comment:
And I would just note that this issue with `dladdr-6` is not resolved yet. As I noted in past, there can be a different symbol prior `dladdr` and therefore this test cannot check there there is no symbol. I rather proposed to remove this test as we do not know what should we expect from `dladdr-6` check, it is undefined.

Reply:
I'm still looking for a way to add a more reliable way to test the given code path. It is required to use an address, which is not pointing to an import thunk and not pointing to an address having a symbol.

Reply:
I understand what you are trying... What about passing here base address of some library (as returned by dli_fbase field of dladdr())? Base address should not contain any symbol and is part of mapped library.

Reply:
I choosed main-6 which is sure to use, as there is code before

Reply:
But `main-6` has same problem as before. You put into test that is has no symbol, but there can be other function with exported symbol on that address. There is nothing which says that on Linux address `main-6` cannot be other exported symbol and therefore this test would fail.